File: python/Problem89.py
# Problem: 89. Gray Code
# URL: https://leetcode.com/problems/gray-code/
# Difficulty: Medium
# Tags: Bitwise Operation

from typing import List

class Solution:
  def grayCode(self, n: int) -> List[int]:
    results: List[int] = []
    grayCode = 0
    for i in range(2 << (n - 1)):
      grayCode ^= i & (-i)
      results.append(grayCode)
    return results

File: cpp/Problem89.cpp
/*
 * Problem: 89. Gray Code
 * URL: https://leetcode.com/problems/gray-code/
 * Difficulty: Medium
 * Tags: Bitwise Operation
 */

#include <vector>

using namespace std;

class Solution {
public:
	vector<int> grayCode(int n) {
		vector<int> results;
		int grayCode = 0;
		for (int i = 0; i < 2 << (n-1); i++) {
			grayCode ^= i & (-i);
			results.push_back(grayCode);
		}
		return results;
	}
};